Get WebKit MALLOC_HEAP_BREAKDOWN Configuration Building Again

stwrt opened the pull request:
#### 33b383131d5edd4552336cfb67114e4dddaaf363
<pre>
Get WebKit MALLOC_HEAP_BREAKDOWN Configuration Building Again
<a href="https://bugs.webkit.org/show_bug.cgi?id=259786">https://bugs.webkit.org/show_bug.cgi?id=259786</a>

Reviewed by Elliott Williams.

Get the Malloc Heap Breakdown build compiling again. Lots of header includes and exporting of symbols.
This really should not touch the system in any other way, but it does modify files across numerous components.

* Source/JavaScriptCore/assembler/LinkBuffer.cpp:
* Source/JavaScriptCore/assembler/LinkBuffer.h:
* Source/JavaScriptCore/bytecode/InstructionStream.cpp:
* Source/JavaScriptCore/runtime/StringReplaceCache.h:
* Source/JavaScriptCore/runtime/StringSplitCache.h:
* Source/WTF/wtf/BitVector.cpp:
* Source/WTF/wtf/DebugHeap.h:
* Source/WTF/wtf/FastMalloc.h:
* Source/WTF/wtf/MetaAllocator.cpp:
* Source/WTF/wtf/text/CString.h:
* Source/WTF/wtf/text/cf/StringImplCF.cpp:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUBindGroupLayout.h:
* Source/WebGPU/WGSL/AST/ASTBuilder.h:
* Source/WebKit/Shared/ShareableBitmap.cpp:
* Source/WebKit/Shared/ShareableBitmap.h:
* Source/WebKit/Shared/ShareableBitmapHandle.cpp:
* Source/WebKit/UIProcess/mac/LegacySessionStateCoding.cpp:

Canonical link: <a href="https://commits.webkit.org/266981@main">https://commits.webkit.org/266981@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d92b7914c96bb7500b12259258804912cb26435e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15290 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15593 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15955 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17046 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14352 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15430 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/18110 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15693 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/16960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15473 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15903 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13002 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17779 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13182 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13795 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/20747 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/13104 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14263 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13962 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17222 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/14537 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14531 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/12311 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/15483 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13801 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3940 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3671 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18148 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/15718 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14364 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3752 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rreno commented on Source/JavaScriptCore/assembler/LinkBuffer.h:
```diff
@@ -39,10 +39,13 @@
 
 namespace JSC {
 
+DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(BranchCompactionLinkBuffer);
```
This should stay in the .cpp since the class is wholly contained there.

stwrt replied:
Get this error if I keep I keep it in the cpp file.

```
JavaScriptCore GenerateTAPI JavaScriptCore.tbd
warning: violations found for arm64e
error: no declaration found for exported symbol 'JSC::BranchCompactionLinkBufferMalloc::debugHeap()' in dynamic library
WebGPU CompileC Adapter.o
WebGPU CompileC CommandEncoder.o
```

emw-apple replied:
If it stays in the class, it needs to have visibility hidden, because otherwise you're exporting the debugHeap() symbol without putting a declaration in headers for a client to import.

rreno replied:
Yep so you should pass `WTF_PRIVATE` as the second parameter to the DECLARE macro.

rreno replied:
jinx

stwrt replied:
Changed to `DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER_AND_EXPORT(BranchCompactionLinkBuffer, WTF_INTERNAL);
` and it builds successfully now.

rreno commented on Source/JavaScriptCore/runtime/StringReplaceCache.h:
```diff
@@ -30,14 +30,11 @@
 class StringReplaceCache {
-    WTF_MAKE_FAST_ALLOCATED;
-    WTF_MAKE_NONCOPYABLE(StringReplaceCache);
+
```
This seems unnecessary.

stwrt replied:
String Replace Cache was having some issues with compiling; need to look into the again. I noticed some of the other similar caches did not use the FAST_ALLOCATED, so I removed this.

stwrt replied:
Re-added. Was able to resolve with `DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER_AND_EXPORT`.

rreno commented on Source/JavaScriptCore/runtime/StringSplitCache.h:
```diff
@@ -30,13 +30,9 @@
 class JSImmutableButterfly;
 
 class StringSplitCache {
-    WTF_MAKE_FAST_ALLOCATED;
```
Ditto StringReplaceCache. Why are you removing the noncopyable/fast malloc macros?

justinmichaud replied:
Worth noting: this will almost certainly have perf implications

stwrt replied:
Added back.

rreno commented on Source/WTF/wtf/BitVector.h:
```diff
@@ -45,6 +45,8 @@ namespace WTF {
 
 class FixedBitVector;
 
+DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(BitVector);
```
You didn't remove the WTF_MAKE_FAST_ALLOCATED from the class definition.

stwrt replied:
Where is WTF_MAKE_FAST_ALLOCATED in the .cpp file?

rreno commented on Source/WTF/wtf/MetaAllocator.h:
```diff
@@ -41,6 +41,8 @@ namespace WTF {
 
 #define ENABLE_META_ALLOCATOR_PROFILE 0
 
+DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(MetaAllocatorFreeSpace);
```
This should probably have the same name as the nested class (i.e. `MetaAllocatorFreeSpaceNode`).

rreno replied:
I understand that it didn't before your patch.

stwrt replied:
This can be handled in a future patch then. It falls outside the current patch's scope.

rreno commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -63,6 +63,9 @@ class LLIntOffsetsExtractor;
 
 namespace WTF {
 
+DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(StringWrapperCFAllocator);
```
Why did this need to move to the header?

rreno commented on Source/WTF/wtf/text/cf/StringImplCF.cpp:
```diff
@@ -32,10 +32,9 @@
 namespace WTF {
 
-namespace StringWrapperCFAllocator {
+DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER(StringWrapperCFAllocator);
```
Why did this move out of the namespace?

rreno commented on Source/WebGPU/WGSL/AST/ASTBuilder.h:
```diff
@@ -46,10 +48,10 @@ class ShaderModule;
 
-DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(WGSLAST);
 
+class Node;
```
Unnecessary change.

rreno commented on Source/WebKit/Shared/ShareableBitmap.h:
```diff
@@ -116,7 +119,10 @@ class ShareableBitmapHandle  {
+DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER_AND_EXPORT(ShareableBitmap, WTF_INTERNAL);
+
 class ShareableBitmap : public ThreadSafeRefCounted<ShareableBitmap> {
+    WTF_MAKE_FAST_ALLOCATED;
```
Why this and not _WITH_HEAP_IDENTIFIER?

rreno commented on Source/WebKit/Shared/ShareableBitmap.h:
```diff
@@ -93,6 +95,7 @@ class ShareableBitmapConfiguration {
 };
 
 class ShareableBitmapHandle  {
+    WTF_MAKE_FAST_ALLOCATED;
```
Why?

rreno commented on Source/WTF/wtf/DebugHeap.h:
```diff
@@ -55,9 +55,11 @@ class DebugHeap {
-#define DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(Type) \
+#define DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(Type) DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER_AND_EXPORT(Type, WTF_EXPORT_PRIVATE)
+
+#define DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER_AND_EXPORT(Type, Export) \
```
You'll also need this macro available in non MHB builds.

stwrt replied:
Right. Need to add support for the non MHB #define

emw-apple commented on Source/JavaScriptCore/assembler/LinkBuffer.h:
```diff
@@ -39,6 +39,7 @@
 #include "MacroAssemblerCodeRef.h"
 #include <wtf/DataLog.h>
 #include <wtf/FastMalloc.h>
+#include <wtf/NeverDestroyed.h>
```
Seems like this include should be in `DebugHeap.h`, since that's where the macro is defined.

emw-apple commented on Source/JavaScriptCore/bytecode/InstructionStream.cpp:
```diff
@@ -26,6 +26,8 @@
 #include "config.h"
 #include "InstructionStream.h"
 
+#include <wtf/NeverDestroyed.h>
```
Ditto.

stwrt replied:
```
JavaScriptCore CompileC UnifiedSource33.o
JavaScriptCore CompileC UnifiedSource32.o
In file included from /Users/bts/dev/OpenSource/WebKitBuild/Release/DerivedSources/JavaScriptCore/unified-sources/UnifiedSource32.cpp:1:
./bytecode/InstructionStream.cpp:33:1: error: implicit instantiation of undefined template 'WTF::LazyNeverDestroyed<WTF::DebugHeap>'
DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER(InstructionStream);
^
In file included from /Users/bts/dev/OpenSource/WebKitBuild/Release/DerivedSources/JavaScriptCore/unified-sources/UnifiedSource32.cpp:1:
In file included from ./bytecode/InstructionStream.cpp:26:
In file included from /Users/bts/dev/OpenSource/Source/JavaScriptCore/config.h:39:
In file included from /Users/bts/dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/FastMalloc.h:25:
/Users/bts/dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/DebugHeap.h:82:51: note: expanded from macro 'DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER'
        static LazyNeverDestroyed<WTF::DebugHeap> heap; \
In file included from /Users/bts/dev/OpenSource/WebKitBuild/Release/DerivedSources/JavaScriptCore/unified-sources/UnifiedSource32.cpp:1:
In file included from ./bytecode/InstructionStream.cpp:27:
In file included from ./bytecode/InstructionStream.h:29:
In file included from ./bytecode/BytecodeIndex.h:28:
In file included from /Users/bts/dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/HashTraits.h:25:
/Users/bts/dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/Forward.h:78:61: note: template is declared here
template<typename, typename = AnyThreadsAccessTraits> class LazyNeverDestroyed;
1 error generated.
```

emw-apple commented on Source/JavaScriptCore/runtime/StringSplitCache.h:
```diff
@@ -27,10 +27,11 @@
 #pragma once
 
 #include <array>
+#include <wtf/FastMalloc.h>
```
Ditto.

emw-apple commented on Source/WTF/wtf/DebugHeap.h:
```diff
@@ -88,7 +90,9 @@ class DebugHeap {
 #else // ENABLE(MALLOC_HEAP_BREAKDOWN)
 
-#define DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(Type) \
+#define DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(Type) DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER_AND_EXPORT(Type, nullptr)
```
`nullptr` is weird, since this parameter is never used like a pointer. I think I would just put WTF_EXPORT_PRIVATE here? (That would also enable you to hoist this definition outside of the `#ifdef` and only define it once, if you'd like.)

emw-apple commented on Source/WTF/wtf/text/CString.h:
```diff
@@ -26,6 +26,7 @@
 #pragma once
 
 #include <span>
+#include <wtf/FastMalloc.h>
```
Ditto.

emw-apple commented on Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUBindGroupLayout.h:
```diff
@@ -25,6 +25,7 @@
 
 #pragma once
 
+#include <wtf/FastMalloc.h>
```
Ditto.

emw-apple commented on Source/WebGPU/WGSL/AST/ASTBuilder.h:
```diff
@@ -25,7 +25,9 @@
 
+#include <wtf/FastMalloc.h>
 #include <wtf/MallocPtr.h>
+#include <wtf/NeverDestroyed.h>
```
Ditto.

stwrt replied:
```
WGSL CompileC ASTBuilder.o
/Users/bts/dev/OpenSource/Source/WebGPU/WGSL/AST/ASTBuilder.cpp:33:1: error: implicit instantiation of undefined template 'WTF::LazyNeverDestroyed<WTF::DebugHeap>'
DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER(WGSLAST);
^
In file included from /Users/bts/dev/OpenSource/Source/WebGPU/WGSL/AST/ASTBuilder.cpp:27:
In file included from /Users/bts/dev/OpenSource/Source/WebGPU/WGSL/AST/ASTBuilder.h:28:
/Users/bts/dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/DebugHeap.h:82:51: note: expanded from macro 'DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER'
        static LazyNeverDestroyed<WTF::DebugHeap> heap; \
In file included from /Users/bts/dev/OpenSource/Source/WebGPU/WGSL/AST/ASTBuilder.cpp:27:
In file included from /Users/bts/dev/OpenSource/Source/WebGPU/WGSL/AST/ASTBuilder.h:32:
In file included from /Users/bts/dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/Vector.h:33:
/Users/bts/dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/Forward.h:78:61: note: template is declared here
template<typename, typename = AnyThreadsAccessTraits> class LazyNeverDestroyed;
1 error generated.
```

emw-apple commented on Source/WebKit/Shared/ShareableBitmap.h:
```diff
@@ -31,6 +31,8 @@
 #include <WebCore/PlatformImage.h>
 #include <wtf/ArgumentCoder.h>
+#include <wtf/ExportMacros.h>
+#include <wtf/FastMalloc.h>
```
Ditto.

emw-apple reviewed (commented):
The proliferation of `WTF_INTERNAL` outside of WTF is unfortunate. But I don't think we have a good alternative, unless we were to refactor `DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER` to **not** default to public visibility. Please feel free to do that, but I'm not going to block the PR on it.

Otherwise, looks good to me except for a few nits:

stwrt committed:
Get WebKit MALLOC_HEAP_BREAKDOWN Configuration Building Again
https://bugs.webkit.org/show_bug.cgi?id=259786

Reviewed by Elliott Williams.

Get the Malloc Heap Breakdown build compiling again. Lots of header includes and exporting of symbols.
This really should not touch the system in any other way, but it does modify files across numerous components.

* Source/JavaScriptCore/assembler/LinkBuffer.cpp:
* Source/JavaScriptCore/assembler/LinkBuffer.h:
* Source/JavaScriptCore/bytecode/InstructionStream.cpp:
* Source/JavaScriptCore/runtime/StringReplaceCache.h:
* Source/JavaScriptCore/runtime/StringSplitCache.h:
* Source/WTF/wtf/BitVector.cpp:
* Source/WTF/wtf/DebugHeap.h:
* Source/WTF/wtf/FastMalloc.h:
* Source/WTF/wtf/MetaAllocator.cpp:
* Source/WTF/wtf/text/CString.h:
* Source/WTF/wtf/text/cf/StringImplCF.cpp:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUBindGroupLayout.h:
* Source/WebGPU/WGSL/AST/ASTBuilder.h:
* Source/WebKit/Shared/ShareableBitmap.cpp:
* Source/WebKit/Shared/ShareableBitmap.h:
* Source/WebKit/Shared/ShareableBitmapHandle.cpp:
* Source/WebKit/UIProcess/mac/LegacySessionStateCoding.cpp:

Canonical link: https://commits.webkit.org/266981@main
